src/main/scala/is/hail/expr/Type.scala

```diff
@@ -1862,7 +1871,10 @@ final case class TStruct(fields: IndexedSeq[Field], override val required: Boole
             rvb.startStruct()
             var i = 0
             while (i < localSize) {
-              rvb.addField(this, region, offset, i)
+              if (region != null)
```

Comment:
Is this the only necessary check?

I admittedly don't fully understand the invariants here.

The way I'm thinking is: `region == null` means that we are inside a part of the type tree that didn't exist previously. We need to know this because we shouldn't try to copy fields from a non-existing struct.

In this portion of the type tree, the `this` object being inserted into is always an empty struct, so we always hit the `None` case of the `selfField(key)` match statement.

Reply:
I think so.

 > The way I'm thinking is: region == null means that we are inside a part of the type tree that didn't exist previously. We need to know this because we shouldn't try to copy fields from a non-existing struct.

Correct.

 > In this portion of the type tree, the this object being inserted into is always an empty struct

What?  No.  How do you figure?  This is TStruct.  Can can be any pre-existing struct we're inserting into.

Reply:
Ah OK, I completely misunderstood what `region == null` meant. I thought it meant the type we were currently inserting into was _statically_ absent.

Maybe I'm dense, but I feel like this code would be easier to understand if we passed an explicit `missing` boolean and collapsed the correlated conditionals. I've been using the `offset, missing` pair paradigm in the compiler pretty extensively (it's true this wastes a whole bit since we `region == null` is back to an invalid state, but I like the self-documenting nature of the parameter named missing).

```scala
(updateKey(key, j, insertedFieldType), { (region, offset, missing, rvb, inserter) =>
  if (missing) {
    var i = 0
    while (i < localSize) {
      if (i == j) fieldInserter(region, 0, true, rvb, inserter)
      else rvb.setMissing()
      i += 1
    }
  } else {
    rvb.startStruct()
    var i = 0
    while (i < j) {
      rvb.addField(this, region, offset, i)
      i += 1
    }
    fieldInserter(region, loadField(region, offset, j), rvb, inserter)
    i += 1
    while (i < localSize) {
      rvb.addField(this, region, offset, i)
      i += 1
    }
    rvb.endStruct()
  }
}
```

Reply:
I guess it's fine as you've written it too. This region value code is quite different from normal scala and I often find it difficult to parse out the meaning on first read.